### PR TITLE
feat(dynamic-forms): adds warning messages

### PR DESCRIPTION
**Pull request by mario-subo**

### CODE REVIEW

File: lib/src/hooks/useDynamicForm.tsx [hunk 1]
```diff
@@ -91,73 +94,83 @@ type OnChangeFormItem = (item: DynamicFormItem) => OnChangeInput;
...
-            let initialValue: any;
+export type FormRenderFunction = () => JSX.Element;
+export type SetDynamicFormState = React.Dispatch<React.SetStateAction<DynamicFormInternalState>>;
+export type SetDynamicFormErrors = React.Dispatch<React.SetStateAction<DynamicFormErrors>>;
+export type SetDynamicFormWarnings = React.Dispatch<React.SetStateAction<DynamicFormErrors>>;
+type isDynamicFormDirty = boolean;
+export type UseDynamicForm = [FormRenderFunction, DynamicFormInternalState, SetDynamicFormState, SetDynamicFormErrors, SetDynamicFormWarnings, DynamicFormMetaData, isDynamicFormDirty];
+export function useDynamicForm(sections: DynamicFormSection[]): UseDynamicForm {
+    const initialState: DynamicFormInternalState = useMemo(() => {
+        let initialFormState: DynamicFormInternalState = {};
```
Comment (eweseong) [inline lib/src/hooks/useDynamicForm.tsx:105]: notice there's no reference changes, perhaps we can use `const initialFormState`

File: lib/src/hooks/useDynamicForm.tsx [hunk 2]
```diff
@@ -91,73 +94,83 @@ type OnChangeFormItem = (item: DynamicFormItem) => OnChangeInput;
...
+        sections?.map((section) => {
```
Comment (omarVodiaK) [inline lib/src/hooks/useDynamicForm.tsx:106]:
```suggestion
        sections?.map((section: DynamicFormSection) => {
```

File: lib/src/hooks/useDynamicForm.tsx [hunk 3]
```diff
@@ -91,73 +94,83 @@ type OnChangeFormItem = (item: DynamicFormItem) => OnChangeInput;
...
+            initialFormState[section?.key] = {};
+            section.items?.map((item) => {
```
Comment (omarVodiaK) [inline lib/src/hooks/useDynamicForm.tsx:108]:
```suggestion
            section.items?.map((item: DynamicFormItem) => {
```

File: lib/src/hooks/useDynamicForm.tsx [hunk 4]
```diff
@@ -91,73 +94,83 @@ type OnChangeFormItem = (item: DynamicFormItem) => OnChangeInput;
...
 
-            switch (controlType) {
-                case "Dropdown": {
-                    if (multi) {
-                        if (value !== null && value !== undefined && Array.isArray(value) && (value as any[]).every((x) => typeof x === "string")) {
-                            initialValue = value as string[];
+                switch (controlType) {
+                    case "Dropdown": {
+                        if (multi) {
+                            if (value !== null && value !== undefined && Array.isArray(value) && (value as any[]).every((x) => typeof x === "string")) {
```
Comment (omarVodiaK) [inline lib/src/hooks/useDynamicForm.tsx:115]: if value is an empty array this line will return true, if this is the correct use case then i suggest to use:
```suggestion
                            if (!!(value ?? (Array.isArray(value) && (value as any[]).every((x) => typeof x === "string")))) {
```

**Review by omarVodiaK — APPROVED** [review]

**Review by eweseong — APPROVED** [review]

File: lib/src/hooks/useDynamicForm.tsx (hunk 4 above)
Comment (eweseong) [reply lib/src/hooks/useDynamicForm.tsx:115]: on top of Omar's comment, could we let the `.isArray` to do the checking for `null` and `undefined`?

```
if (Array.isArray(value) && (value as any[]).every((x) => typeof x === "string")) {
```

Comment (mario-subo) [reply lib/src/hooks/useDynamicForm.tsx:115]: Omar it's ok to evaluate to true if array is empty as it's a valid state. II will go with Yeoh's comment, I agree the null and undefined check is not needed as will be done is isArray.

**Pull request merged**